tests/integration/datastore/test_datastore.py

```diff
@@ -14,64 +14,392 @@
+        should_error = test_record.test
 
-    # Test get PolicyArrangement
-    query_arrangement = test_datastore.get_policy_arrangement(arrangement_id)
-    assert new_arrangement == query_arrangement
+    # Records can also have ints as IDs
+    with storage.describe(TestRecord, 1337, writeable=True) as test_record:
+        test_record.test = b'test int ID'
+        assert test_record.test == b'test int ID'
 
-    # Test del PolicyArrangement
-    test_datastore.del_policy_arrangement(arrangement_id)
-    with pytest.raises(datastore.NotFound):
-        del_key = test_datastore.get_policy_arrangement(arrangement_id)
+    # Writing to a non-existent field errors
+    with pytest.raises(datastore.DatastoreTransactionError):
+        with storage.describe(TestRecord, 'test_id', writeable=True) as test_record:
+            test_record.nonexistent_field = b'this will error'
 
+    # Writing the wrong type to a field errors
+    with pytest.raises(datastore.DatastoreTransactionError):
+        with storage.describe(TestRecord, 'test_id', writeable=True) as test_record:
+            test_record.test = 1234
 
-def test_workorder_sqlite_datastore(test_datastore):
-    bob_keypair_sig1 = keypairs.SigningKeypair(generate_keys_if_needed=True)
-    bob_keypair_sig2 = keypairs.SigningKeypair(generate_keys_if_needed=True)
+    # Check that nothing was written
+    with storage.describe(TestRecord, 'test_id') as test_record:
+        assert test_record.test != 1234
 
-    arrangement_id = b'test'
+    # Any unhandled errors in the context manager results in a transaction abort
+    with pytest.raises(datastore.DatastoreTransactionError):
+        with storage.describe(TestRecord, 'test_id', writeable=True) as test_record:
+            # Valid write
+            test_record.test = b'this will not persist'
+            # Erroneous write causing an abort
+            test_record.nonexistent = b'causes an error and aborts the write'
 
-    # Test add workorder
-    new_workorder1 = test_datastore.save_workorder(bob_keypair_sig1.pubkey, b'test0', arrangement_id)
-    new_workorder2 = test_datastore.save_workorder(bob_keypair_sig2.pubkey, b'test1', arrangement_id)
+    # Check that nothing was written from the aborted transaction above.
+    with storage.describe(TestRecord, 'test_id') as test_record:
+        assert test_record.test == b'test data'
 
-    # Test get workorder
-    query_workorders = test_datastore.get_workorders(arrangement_id)
-    assert {new_workorder1, new_workorder2}.issubset(query_workorders)
+    # However, a handled error will not cause an abort.
+    with storage.describe(TestRecord, 'test_id', writeable=True) as test_record:
+        # Valid operation
+        test_record.test = b'this will persist'
+        try:
+            # Maybe we don't know that this field exists or not?
+            # Erroneous, but handled, operation -- doesn't cause an abort
+            should_error = test_record.bad_read
+        except TypeError:
```

Comment:
Would it be more logical to raise `AttributeError` if the field is not found?

Reply:
Hm, actually just below a similar construction does raise an `AttributeError`...

Reply:
I have very uncertain opinions about how I use these two errors, and it's one of the most obvious sore spots in this code (for me, at least).

The difference between these errors is mostly a Pythonic notion. A `TypeError` is raised when you try to do an operation on an invalid type. For example, `[print(i) for i in 1234]` is a `TypeError` because you can't iterate over an `int`. You can think of it like trying to do an unsupported operation.
In contrast, an `AttributeError` is for errors with assignment or references. The datastore API, unfortunately, has the ill effect of blurring these lines sometimes.

This is a `TypeError` because you're trying to read from a record with a field that doesn't exist in its model definition. It goes beyond an "unsupported operation" and into this just not being possible at all.

The `AttributeError` below is very similar, but it's also very different: you're trying to read from a record that does have this field in its model definition, but has no data. This is indeed a supported operation, meaning you're still able to write to it. In comparison to the `TypeError`, you're not able to write to that field at all (or ever) until the model is updated. Does that make sense?

Reply:
Consider a custom exception class.

Reply:
I actually find Python inconsistent here. `[print(i) for i in 1234]` checks for `__iter()__`, doesn't find it and raises `TypeError`. But if you try `(1234).__iter__()`, you'll get `AttributeError`. Would be nice to have some kind of `InterfaceError`, but alas...

So for me a more expected behavior would be 
- an `AttributeError` if the field doesn't exist in the model
- a `ValueError` if you try to read from a field that exists but has no data (why even raise anything, by the way? Would just returning `None` work?). Perhaps a custom exception would be even better than `ValueError` here.

Reply:
Adding to that, if reading from an existing field with no data is an error, why is it possible to create a record with such a field in the first place?

Reply:
> Adding to that, if reading from an existing field with no data is an error, why is it possible to create a record with such a field in the first place?

Why would you not be able to? This is the equivalent of a "nullable" field in any sort of relational DB. An example we use of this is when we handle arrangements for the first time from Alice; we create an arrangement without a KFrag and set it when it's received. This logic would prevent us from doing that.

> Would just returning `None` work?

Not really. There's a very subtle problem with this logic, and a more practical usability issue. If I return `None` from a field, am I returning `None` or does the field have a `NoneType` in it?
For the usability issue, the argument is much better. Reading an empty field in a record _should not be silent_. It should raise an error at a minimum. This ensures that datastore readers handle null logic, and prevents a `None` from bubbling up through the code where it gets difficult to debug. You're right, though: a custom error type might make more sense for this part.

> Consider a custom exception class.

I intend to, but I need to figure out a more broad direction of where the datastore's API is going to go, and what cases I may figure out. I'd like to simplify this along with some `__setattr__` and `__getattr__` logic with better errors, but it needs iterative work.

Reply:
Additionally, current deletion logic is handled by setting a field to `None`, which tells the datastore to delete the field.